core/src/services/s3/core.rs

```diff
@@ -246,6 +249,22 @@ impl S3Core {
 
         req
     }
+
+    pub fn insert_checksum_header(
+        &self,
+        mut req: http::request::Builder,
+        body: &Buffer,
+    ) -> http::request::Builder {
+        if let Some(checksum_algorithm) = self.checksum_algorithm.as_ref() {
+            let checksum = match checksum_algorithm {
+                S3ChecksumAlgorithm::Crc32c => {
+                    BASE64_STANDARD.encode(crc32c::crc32c(body.to_vec().as_slice()).to_be_bytes())
```

Comment:
Please don't call `body.to_vec()` which invovles extra allocation and copy. Please call `crc32c_append` in loop instead.